### DISCUSSION TITLE
New entity component: `todo`

### DISCUSSION THREAD

**Discussion by allenporter** [home-assistant/architecture#962] [Entity Models] [answered]
# Context

A To-do Item represents an action-item or task to complete with a name such as "Buy coffee" or "Submit income tax return". A To-do item can be incomplete or complete. 

A To-do List is a collection of To-do Items. A To-do list has a name such as "Shopping List" or "My tasks". A To-do list has an ordering of the To-do items on the list.

The existing use of To-do lists in Home Assistant is through integrations that all work in unique ways:
- Shopping list (11,868 installs) has a full end-to-end integration with the frontend and voice and service calls for management. It supports a single list.
- Todoist API (480 installs) supports multiple lists, exposes To-do items as a calendar and a service that allows creating tasks
- Habitica (100 installs) which supports pass through for the full API which includes habits/tasks/to-dos.
- Remember the milk (33 installs) has service calls for manipulating a single list, though the API appears to support multiple.
- There are also custom components such as Grocy, Google Tasks, and possibly others that appear to return large tasks list items as attributes on a sensor.

rfc5545 (our favorite!) has a standard specification of To-do items with many other descriptive properties (description, location, categories, classification), date/time properties like due dates, and recurrence.

There was a previous discussion about a [task domain](https://github.com/home-assistant/architecture/discussions/898) however it did not result in a concrete technical proposal beyond feature requests. As a result, I've started a separate discussion.

### Shopping List

Since `shopping_list` it the most popular To-do item integration and is a fully functioning To-do list, we'll look a little closer at it. The main feature gap is that it does not support multiple lists. Here is a reminder of how it works in more detail:
- There are no entities.
- The services for shopping list items operations are `add_item`, `remove_item`, `complete_item`, and `incomplete_item` which all accept a name field. (A list item has a name and a status only.)
- The services for shopping lists operations are `complete_all`, `incomplete_all`, `clear_completed_items` and `sort`.
- The primary UI use cases supported via Web Socket API are:
  - List Item: Add a new item, Rename the item, Mark complete, Mark incomplete, Reorder an item in the list
  - List: Clear all completed items
- There is an intent for adding an item, and another for listing the first 5 items. I believe these may be marked for deprecation, however, but i am not that familiar with Intents
- The `shopping_list` event type contains details about the actions performed on the shopping list {"add", "update", "update_list", "reorder", "sorted"} and in some cases contains the current status of the shopping list items affected.
- The re-order operation takes as input the entire list of incomplete tasks.

# Proposal
Introduce a `todo` entity component. 

### Goals

The high level goals are:
- Support multiple To-do Lists
- It should be east to add, remove, update, re-order To-do items in a To-do List from the UI
- Automations should be able to add and remove To-do items from a To-do List and update their status

## Todo component
The proposal is to add a `todo` entity component which can be thought of as similar to the calendar component, but with existing functionality of `shopping_list` for each To-do List.

A TodoListEntity supports operations to create, delete, update a `TodoItem` dataclass similar to a calendar where integrations provide the concrete implementations. The entity state is the ISO 8601 timestamp of when the list was last updated. For example: "2023-09-04T22:59:32.360725+00:00"

The properties on a To-do Item supported would be a subset of `VTODO` properties from rfc5545 and I am not aware of more needed. However, we only use a subset in order to support a broad set of To-do List integrations since not all follow that spec.

Reordering needs a new approach. I recommend we follow the Google Tasks approach which supports a move operation that references a previous item. I like this because it is probably the lowest common denominator to implement across all integrations. Not all integrations support task reordering, so this will need to be optional. For example, Todoist has a manual sort order for projects.

### Integration Considerations

- Shopping List, Todoist, can all provide TodoListEntity. I'm not as familiar with Habatica and Remember the Milk which look like they can work but I did not look too closely.
- The local_calendar integration's underlying library already fully supports VTODO components from rfc5545 and all the related properties.I can update local calendar to also support adding To-do Lists in the config flow. (see future considerations below w.r.t calendar)
- The caldav integration's underlying library appears to support VTODO and can be extended to support To-do lists. This can let us support To-do lists from iCloud, Nextcloud, and other providers.
- I'd like to also add a Google Tasks integration. I've verified the cloud APIs are compatible with this proposal.
- I've done some initial looking at other To-do/task related integrations that I do not personally plan to drive forward, but I wanted to make sure they are compatible with this proposal.  Microsoft Team Graph has a Tasks API that I've verified also looks compatible with this proposal. Google Keep also supports Lists which could perhaps use this, but I didn't look too closely.

Thank you to @Joostlek for doing survey of the integrations and recent PRs.

#  Alternatives
- The name `todo` is chosen in line with rfc5545 and since it is very specific. However a previous discussion proposed `task` which is a bit more generic and could get confused with other types of object (e.g. asyncio task).  Task may translate better since "Todo" may be confusing for Spanish users. Happy to defer to tech writers here.
- We could do more complex things with entity state  and attribute such as returning the # of items in the list, however I worry about encoding meaning that will be hard to undo/change later (e.g. `calendar` has a bunch of side effects because it acts like a binary sensor and tracking future items). The timestamp was chosen so that the entity state could potentially serve as a simple way to detect changes like the existing shopping list event.
- Go broader: Support a list entity component. A list could be either read-only or read-write and allow generic use cases of lists of data (outside of the state machine). A List Item could have basic properties that allow emulating a To-do Item. An input_list could let users interactively populate lists.
- Go narrower: Extend shopping lists to support multiple lists. We could assume a default shopping list everywhere and add an optional list parameter to the existing APIs.
- Go narrower: Add another new local To-do list component as a one-off which supports multiple lists, built on top of our calendar library, without making a general API. This could satisfy any voice use cases that require multiple lists.
- Don't generalize at all: Stick with only single list support for shopping list and don't offer any new features here. Any new To-do list integrations can be standalone one-offs.

# Consequences
- There are additional UI considerations to work through: We probably only want a single place to have To-do Lists in the UI, and likely want to extend the existing shopping list UI.
- There are many users of the existing Shopping List APIs. I think for now we should preserve all these APIs and allow them to continue to work, but there may be long term implications I am not yet aware of here that could cause a maintenance burden. This may be confusing to have multiple ways to do the same thing (native Todo APIs vs Shopping List APIs). Some other existing shopping list services like clear_completed_items and complete_all will not be supported without further discussion.
- Considering some of the implementation details, there may be some specific performance concerns to reconcile with the API such as supporting batching. For example, we could support "clear completed items" by sending a bulk delete of all items.
- Home maintenance is a common use case where people think about To-do Lists. In the above propsoal, integrations would not directly read/write Todo items (e.g. "Change sensor battery") and that would be out of scope. However, the proposal can support automations managing To-do Item lifecycle.  In other words, we are not right now considering a case where integrations surface their own todo lists or modify todo list items for maintenance.

# Future Considerations
- There are many additional To-do item properties such as 'description' that we could consider in the future. There are a handful of properties supported by common To-do component implementations (Todoist, Google Tasks, Microsoft Team Graph, etc) and by rfc5545.
- Expanding with even more rfc5545 extensions could allow support for recurring To-do Items. My current thinking is that this could be supported via exposing To-do items through calendar rather than adding native support for recurrence to the To-do component. Local Calendar and CalDAV could possibly support this, for example, and would need to be done per-integration and would not need additional API support.

**Comment by balloob** [comment, marked as answer]
I was nodding along all the time while reading until you got to recurring to-do items 🙈 

I think that we should keep it task related and not expand the scope to list. Because we would want to support some to do attributes (like due date, checked/unchecked) and so it would then be weird why another domain that uses lists (can't think of 1 right now) couldn't add their supported attributes.

It would be cool if these entities could also have the tasks with due dates be shown on the calendar view.

> We could do more complex things with entity state and attribute such as returning the # of items in the list, however I worry about encoding meaning that will be hard to undo/change later (e.g. calendar has a bunch of side effects because it acts like a binary sensor and tracking future items). The timestamp was chosen so that the entity state could potentially serve as a simple way to detect changes like the existing shopping list event.

I _kinda_ would lean to have the state reflect the number of unchecked items. Since it's a todo list entity, it seems a natural fit. I don't worry too much about encoding meaning that we want to undo, as long as we stick to it being just for to do items.

**Reply by allenporter** [reply]
Sounds good, we can go with a count as the state.

Will address due dates etc in the future.  We can consider if it makes to expose todo items as a calendar, or do it with a UI-only approach. (Recurrence will be even later if at all)

From our discussion, will expose a "local todo" integration given the existing shopping serivces wouldn't be able to support multiple shopping lists.
